### PR TITLE
fix(core): passphrase-encrypted key generation/signing under OpenSSL 3.6.x

### PR DESCRIPTION
**Pull request by mike-north**
Refs #75

## Root cause

`generateKeyPair`'s public-key extraction step (`pkey -passin stdin`) and `sign`'s `dgst -passin stdin` step failed on OpenSSL 3.6.2 with:

```
Could not find private key of key from <path>
...UI routines:open_console:unknown ttyget errno value:crypto/ui/ui_openssl.c:457:errno=102
```

Investigation (isolated with standalone repro scripts outside vitest) found the failure is **not** about which file descriptor carries the passphrase. A passphrase piped via a dedicated `fd:3` pipe fails identically to one piped via `stdin`, as long as fd 0 (`stdin`) is *also* a Node `spawn`-created pipe. OpenSSL 3.6.x's passphrase-reading UI routine — used by the `pkey` and `dgst` subcommands, but apparently not `genpkey` — falls back to an interactive console prompt (fatal outside a TTY) whenever fd 0 is that kind of pipe, regardless of where the actual passphrase comes from. Setting fd 0 to `'ignore'` avoids the fallback entirely; the passphrase is then supplied cleanly through an unrelated pipe on fd 3.

(An earlier attempt at this fix used a real file-backed fd 3 while leaving stdin as a pipe — that also failed identically once re-verified outside a single-shot script, which is what led to isolating the true variable: `stdin`'s pipe-ness, not fd 3's file-vs-pipe-ness.)

## Fix

`runOpenSSL()` (`packages/core/src/crypto.ts`) now sets `stdio: ['ignore', 'pipe', 'pipe', 'pipe']` when a passphrase is supplied, and writes the passphrase to the fd-3 stream instead of stdin. All three call sites (`genpkey -pass fd:3`, `pkey -passin fd:3`, `dgst -passin fd:3`) were converted uniformly for consistency, since all three now pass reliably with this approach (confirmed via a 5-iteration stress test across all three during development).

## Acceptance criteria → evidence

- **All 10 named failing tests pass on this machine's OpenSSL (3.6.2)**: confirmed — `pnpm --filter @attest-it/core test` is green (698/698), run 5x in a row with no flakiness.
- **Fix uses fd-based mechanism instead of relying on OpenSSL's console fallback**: done — `fd:3` used for all three invocations; see root-cause note above on why `stdin` also had to become `'ignore'`.
- **No regression to the currently-passing `-pass stdin` key-generation step**: converted to `-pass fd:3` for consistency; verified still passing (existing `should generate an encrypted keypair` test, plus new repeated-invocation regression test).
- **Regression test added**: `packages/core/test/crypto.test.ts` gains two tests — repeated public-key extraction from an encrypted key, and repeated signing with an encrypted key — both reference issue #75 in comments and were verified to fail against the pre-fix code (reverted the fix locally and reran; both new tests failed with the original `ttyget errno` error, confirmed via 12/12 failures instead of the usual 10).
- **CI OpenSSL version documented**: CI runs on `ubuntu-latest`, which is Ubuntu 24.04 as of this PR. Per GitHub's [runner-images Ubuntu 24.04 readme](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md), the installed OpenSSL is **3.0.13-0ubuntu3.11** — well below the 3.6.x branch where this bug appears, so **CI does not currently reproduce the pre-fix bug**. The fix is still necessary for contributors on newer OpenSSL (e.g., macOS Homebrew, which is at 3.6.2 as of this PR) and for whenever Ubuntu's `ubuntu-latest` image eventually ships a newer OpenSSL. Added an `openssl version` diagnostic step to `.github/workflows/ci.yml`'s build job so future drift is visible in CI logs going forward.
- **No behavior change to unencrypted paths**: unencrypted `generateKeyPair`/`sign` calls take the `stdio: ['ignore', 'pipe', 'pipe']` branch (no fd 3), unchanged aside from stdin moving from `'pipe'` to `'ignore'` — stdin was never used for anything except the passphrase in this function, so this has no effect on unencrypted paths. Confirmed by the full 696-test pre-existing suite (now 698 with the two new regression tests) passing, including all non-passphrase `generateKeyPair`/`sign`/`verify` tests.

## Verification

- `pnpm --filter @attest-it/core build && pnpm --filter @attest-it/core test`: 698/698 passing, run 5x for flakiness.
- Full monorepo `pnpm run build && pnpm run check && pnpm run test`: green (core 698, cli 634, github-action 62 — all passing).
- `packages/github-action/dist/*.js` regenerated by the build (github-action bundles core's compiled output directly) and included in this PR.

## Changeset

Patch bump for `@attest-it/core` and `attest-it` (the umbrella package re-exports `@attest-it/core` in full).

## Anomalies encountered

My first fix attempt (writing the passphrase to a real temp file and passing its fd via `stdio[3]`, while leaving `stdio[0]` as `'pipe'`) appeared to work in isolated single-shot Node scripts but **did not** fix the actual failing tests when wired into `crypto.ts` and rerun. Isolating this required a side-by-side comparison of the working script vs. the actual call site, which revealed the working script had used `stdio[0] = 'ignore'`, not `'pipe'` — the real fix, as described above. Flagging this in case it's useful context for anyone touching this code later: the interaction is with `stdin`'s pipe-ness, not with which fd carries the passphrase.

### CODE REVIEW

**Comment by mike-north** [comment]
## Root cause of the `build` job's 2 unhandled ECONNRESET errors

Pinned via code analysis (structural, not just timing): `runOpenSSL()` in `packages/core/src/crypto.ts` decided whether to open an extra stdio pipe on fd 3 for the passphrase using `passphrase !== undefined`, while `generateKeyPair()`/`sign()` decided whether to actually reference that fd in OpenSSL's arguments (`-pass fd:3` / `-passin fd:3`) using `if (passphrase)`.

An empty string (`passphrase: ''`) is `!== undefined` but is falsy, so the two checks disagreed:
- `runOpenSSL` opened a 4th stdio slot (`['ignore', 'pipe', 'pipe', 'pipe']`)
- but the `-pass fd:3` / `-passin fd:3` args were never pushed, so OpenSSL never opens or reads fd 3

That left an unconsumed pipe, written to and `.end()`-ed with no `'error'` listener attached. When the OpenSSL child process exits without ever touching fd 3, the parent-side pipe can raise an `ECONNRESET` — and with no listener, that's an unhandled error. `generateKeyPair` makes exactly two such calls (key generation, then public-key extraction), which is why CI saw 2 unhandled errors, both attributed to the "should return encrypted=false for empty passphrase" test in `packages/core/test/key-provider/filesystem-provider.test.ts` — the only test path that calls `generateKeyPair` with an empty-string passphrase.

This is independent of the fd:3 passphrase fix itself (which works correctly for non-empty passphrases) — it's a pre-existing edge case in `runOpenSSL` that the fd:3 change's extra stdio slot exposed.

I could not force the exact ECONNRESET race to reproduce locally on macOS (likely Linux/macOS pipe-teardown timing differs), so instead of a racy regression test I added one that asserts the actual structural bug: it spies on `child_process.spawn` and confirms no 4th stdio slot is opened when the OpenSSL args don't reference `fd:3`. I verified this test fails against the pre-fix code (`stdio` array has length 4) and passes after the fix.

**Fix:** align both checks on passphrase truthiness (`if (passphrase)` in `runOpenSSL` too), and defensively attach an `'error'` listener on the passphrase stream so a similar edge case can't crash the process again.

Local verification: full `pnpm --filter @attest-it/core test` (699 tests) and full `pnpm test` (all packages) pass with zero unhandled errors across multiple runs; `pnpm build && pnpm check` also pass (api-extractor confirms no public API surface change).

**Comment by mike-north** [comment]
Fixed the `Check` failure: the fd:3 regression test used `as string[]` / `as {...}` type assertions on the spawn mock's call args, which the repo's ESLint config bans (`consistent-type-assertions: never`) — replaced them with runtime type guards (`Array.isArray`, `typeof`, `in`) that narrow the types without assertions. `pnpm check` and `pnpm build` are green.

File: packages/core/src/crypto.ts [hunk 1]
```diff
@@ -85,44 +86,67 @@ interface SpawnResult {
...
+    if (passphrase !== undefined) {
+      const passphraseStream = child.stdio[3]
+      if (!(passphraseStream instanceof Writable)) {
+        reject(new Error('Expected a writable stream for the passphrase file descriptor'))
+        return
+      }
+      // Consistently use a newline terminator, matching prior stdin-based behavior.
+      passphraseStream.write(passphrase + '\n')
+      passphraseStream.end()
     }
```
[the remark this answers is not in the document]
Comment (mike-north) [reply packages/core/src/crypto.ts]: This is already covered — commit fb1e96f added a defensive `'error'` listener on `passphraseStream` (packages/core/src/crypto.ts) that is registered before `.write()`/`.end()` are called, so it catches any `'error'` event those calls emit (e.g. EPIPE if OpenSSL exits early on invalid args). The child's real failure is still surfaced via the existing exit-code/stderr check in the `'close'` handler; this listener only prevents an unhandled `'error'` event from crashing the process. Verified the listener registration precedes both calls and re-ran the full crypto test suite (699 tests passing). Marking this thread resolved as outdated.

File: packages/core/test/crypto.test.ts [hunk 2]
```diff
@@ -627,6 +654,25 @@ describe('crypto', () => {
...
+      it('should sign repeatedly with an encrypted key without a console-fallback error', async () => {
+        for (let i = 0; i < 3; i++) {
+          const signature = await sign({
+            privateKeyPath: privatePath,
+            data: `repeat signing test data ${i.toString()}`,
+            passphrase: testPassphrase,
+          })
+
+          expect(signature).toBeTruthy()
+          expect(() => Buffer.from(signature, 'base64')).not.toThrow()
```
[the remark this answers is not in the document]
Comment (mike-north) [reply packages/core/test/crypto.test.ts]: Good catch — fixed in d11d261. Added an `expectValidRsaSignature()` helper in crypto.test.ts that checks the signature string against the base64 alphabet, decodes it and asserts the decoded length equals 256 bytes (the fixed size of an RSA-2048 PKCS#1 v1.5 signature, matching the `rsa_keygen_bits:2048` used by `generateKeyPair`), and confirms re-encoding the decoded bytes reproduces the original string. Replaced all three occurrences of the weak `expect(() => Buffer.from(signature, 'base64')).not.toThrow()` pattern in the file (including this one at the former line 724) with the new helper. Full core test suite (699 tests) still passes.
